crypto/evp/evp_lib.c

```diff
@@ -700,3 +701,130 @@ int EVP_hex2ctrl(int (*cb)(void *ctx, int cmd, void *buf, size_t buflen),
     OPENSSL_free(bin);
     return rv;
 }
+
+#ifndef FIPS_MODE
+/*
+ * TODO(3.0): Temporarily unavailable in FIPS mode. This will need to be added
+ * in later.
+ */
+
+#define MAX_PARAMS 10
+typedef struct {
+    /* Number of the current param */
+    size_t curr;
+    /* Key for the current param */
+    const char *key[MAX_PARAMS];
+    /* Value for the current param */
+    const BIGNUM *bnparam[MAX_PARAMS];
+    /* Size of the buffer required for the BN */
+    size_t bufsz[MAX_PARAMS];
+    /* Running count of the total size required */
+    size_t totsz;
+    int ispublic;
+} PARAMS_TEMPLATE;
+
+static int push_param_bn(PARAMS_TEMPLATE *tmpl, const char *key,
+                         const BIGNUM *bn)
+{
+    int sz;
+
+    sz = BN_num_bytes(bn);
+    if (sz <= 0)
+        return 0;
+    tmpl->key[tmpl->curr] = key;
+    tmpl->bnparam[tmpl->curr] = bn;
+    tmpl->bufsz[tmpl->curr++] = (size_t)sz;
+    tmpl->totsz += sizeof(OSSL_PARAM) + (size_t)sz;
+
+    return 1;
+}
+
+static OSSL_PARAM *param_template_to_param(PARAMS_TEMPLATE *tmpl, size_t *sz)
```

Comment:
Is this worth being a separate PR?

BTW: I like this approach.  The native types might stretch it a bit to handle things going out of scoping.

Reply:
I originally wrote this as a means to an end. If we think this is good general approach then I can have a go a pulling this out as a separate PR and making it more generic.

Reply:
I'm willing to have a go at the other types once this is merged (or before).  I coded something which addresses the alignment issues in #9246.  That should be reusable.

The other thing I think that could be addressed is scoping -- it would be nice if a function could be called that added an auto variable to a template without losing the value.  This shouldn't be too difficult to support.  For the two string types and BIGNUM, it should also be possible (and the _PTR versions not cloning the value).

@levitte what do you think about this approach compared to #9246 and #9254 ?  To me, it seems to sit nicely in the middle.

Reply:
> I'm willing to have a go at the other types once this is merged

Great. In that case. I'll leave this within this PR, and let you refactor it as part of a wider PR.

Reply:
I see this a a temporary measure until we have better key management (i.e. where non-legacy keys are entirely managed in the provider and treated as an object in its own right rather than just a bunch or parameters passed via init).

Reply:
> I see this a a temporary measure until we have better key management (i.e. where non-legacy keys are entirely managed in the provider and treated as an object in its own right rather than just a bunch or parameters passed via init).

Even with the better key management in place we will still need this. For example some keys will be loaded from file by libcrypto and then transferred to the provider. Or similarly keys will be exported from a provider. Or transferred from one provider to another. Some providers may not implement import/export, but I expect at least *our* providers will.

Reply:
I guess we'll have to see.

Reply:
Refer #9305 for a first stab at something similar but including other types.